### ISSUE TITLE
Support adding and removing many relations nicely

### ISSUE THREAD

**Issue by banks** [jonathangeiger/kohana-jelly#4]
assigned to : banks

**Comment by banks** [comment]
I'm thinking about this (out loud). It is actually quite hard to work out how this should work with the current distribution of logic into the field models.

Part of the problem is the age old issue of wether n:many relations should allow adding further restrictions too (they currently do as they return the non-executed model object). If so it is hard to know how to handle relationships that have been assigned but not yet saved.

One solution would be to make the magically accessed relationships always return whole result sets with lazy loading (effectively) so

```
$user->posts
```

Would return a database result object of all related posts. To retrieve a filtered list you would use something like:

```
$user->get('posts')->where(...)
```

so the `get()` method when used directly would return as it does now.

This would allow results of magically accessed relationships to be cached which would increase performance most of the time but would also allow adding and manipulating the related objects prior to saving the model (and all related models).

It would be understood that get() access will never take into account local (unsaved) changes to relationships, but magic access would.

This seems to be the cleanest way to resolve these things to me as it stands.

The has() method would need to work based on the in memory relations rather than database ones though so will need a re-think.

I think we need an interface called Jelly_Field_Relationship with defines a has() method. All relationship types should implement this (even belongsTo). hasOne and belongs to should also implement Jelly_Field_Joinable which defines a with() method. Then in the model code we just need to do instanceof to detect if has/with can be called. I think even though it doesn't necessarily seem right read out that in a 1:1 relationship it is often not important where the foreign key lies and so to have has() work in one direction and not the other seems strange. For belongsTo, has() should still return true if their is a non-null foreign key and it matches.

This relates to the separate with issue.

This has helped me think about the best way to implement this. Jon, please contribute if you have any thoughts. I will start looking at how best to go about this and writing some code.

**Comment by jonathangeiger** [comment]
This all sounds good. How would you like to split up the work? I'll take has() and with()?

There are some questions as to what happens once add and remove are implemented. For example, what happens if you add() or remove() before knowing the results of the initial relationship? We can worry about that later.

I would like to get the API stable as soon as possible. So this should probably be top priority for now.

**Comment by banks** [comment]
> For example, what happens if you add() or remove() before knowing the results of the initial relationship?

If the relationship isn't loaded when calling add() or remove() you have to load it.

Can you implement the has() and with() interfaces in your With branch. Chances are I will have to change the internal API slightly to make the relationship stuff work nicely so if you do that then I will pull your changes to my branch and work with them.

**Comment by jonathangeiger** [comment]
It's up. I've left the belongs_to has() method blank for now, since that will probably get changed by your work.

http://github.com/jonathangeiger/kohana-jelly/commit/18864ea5ca9b6ee0f4023ec751d1a0b4e5b75c7e

**Comment by banks** [comment]
OK, I note you haven't actually added the interfaces is that a miss from your commit or do you want me to add those?

Hopefully I will get some time to get started on this this weekend.

P.S. I discovered yesterday that if you just paste the sha1 of a commit into any message here it will auto link it to the commit and abbreviate it. Cool.

**Comment by banks** [comment]
I forgot to mention, looks good!

**Comment by banks** [comment]
Also, now we have the interfaces, I feel relationship fields should be identified by `instanceof Jelly_Field_Relationship` rather than `in_db = TRUE` because there are cases I can think of where in_db would be true that shouldn't be treated like relationships - password confirmation fields would be one example.

**Comment by jonathangeiger** [comment]
> OK, I note you haven't actually added the interfaces is that a miss from your commit or do you want me to add those?

804cc356368b2b413169b1ceff96208734e51e52

> Also, now we have the interfaces, I feel relationship fields should be identified by instanceof Jelly_Field_Relationship rather than in_db = TRUE because there are cases I can think of where in_db would be true that shouldn't be treated like relationships - password confirmation fields would be one example.

Fixed in aa83933552c9b061519da423bb3690ac9d15bf24 and 1f200182535432f6e42f474747125088515f1bd1

**Comment by banks** [comment]
I posted a message last night but it is not here now... oh well

Thanks for this Jon - you are so quick it makes me look like a slacker!

I will try and get my part of this done as soon as I can - I unfortunately don't have that much time at the moment bu hopefully by this weekend I can have it mostly done.

I mentioned here last night (which is why i posted what I did in the forums) that I feel testing is really crucial now. I think even before the API is 100% stable we need to be thinking of test cases and writing them to be sure that we are fixing bugs not just creating new ones. Something as complex as Jelly really needs thorough testing and I'd be reluctant to say it is a 'stable' release for a while until we have thorough tests and some real-world usage!

My recommendation is we create a separate module called jelly-test which contains all the test code and models and fixtures etc. Kohana's unittest module is now PHPUnit based and is pretty good. Are you comfortable starting on this while I finish up relationships?
